tests/test_messages_queue.cc

```diff
@@ -0,0 +1,482 @@
+  bool triggered_;
+  ProcessingResult result_;
+};
+
+static const std::string kTestMessage = "Test Message";
+static const std::string kTestWorkerMessage = "I am worker thread!";
+
+// Executed on the WorkingThread.
+static void FinishedCallback(ProcessingResult result, FinishTask & finish_task) {
+  // Pass callback result to the future.
+  finish_task(result);
+}
+
+void Test_MessagesQueue_InMemory_Empty() {
+  bool processor_was_called = false;
+  MessagesQueue q;
+  FinishTask finish_task;
+  q.ProcessArchivedFiles([&processor_was_called](bool, const std::string &) {
+    processor_was_called = true;  // This code should not be executed.
+    return false;
+  }, std::bind(&FinishedCallback, std::placeholders::_1, std::ref(finish_task)));
+  TEST_EQUAL(ProcessingResult::ENothingToProcess, finish_task.get());
+  TEST_EQUAL(false, processor_was_called);
+}
+
+void Test_MessagesQueue_InMemory_SuccessfulProcessing() {
+  MessagesQueue q;
+  q.PushMessage(kTestMessage);
+  std::thread worker([&q]() { q.PushMessage(kTestWorkerMessage); });
+  worker.join();
+  bool processor_was_called = false;
+  FinishTask finish_task;
+  q.ProcessArchivedFiles([&processor_was_called](bool is_file, const std::string & messages) {
+    TEST_EQUAL(false, is_file);
+    TEST_EQUAL(messages, kTestMessage + kTestWorkerMessage);
+    processor_was_called = true;
+    return true;
+  }, std::bind(&FinishedCallback, std::placeholders::_1, std::ref(finish_task)));
+  TEST_EQUAL(ProcessingResult::EProcessedSuccessfully, finish_task.get());
+  TEST_EQUAL(true, processor_was_called);
+}
+
+void Test_MessagesQueue_InMemory_FailedProcessing() {
+  MessagesQueue q;
+  q.PushMessage(kTestMessage);
+  bool processor_was_called = false;
+  FinishTask finish_task;
+  q.ProcessArchivedFiles([&processor_was_called](bool is_file, const std::string & messages) {
+    TEST_EQUAL(false, is_file);
+    TEST_EQUAL(messages, kTestMessage);
+    processor_was_called = true;
+    return false;
+  }, std::bind(&FinishedCallback, std::placeholders::_1, std::ref(finish_task)));
+  TEST_EQUAL(ProcessingResult::EProcessingError, finish_task.get());
+  TEST_EQUAL(true, processor_was_called);
+}
+
+void Test_MessagesQueue_SwitchFromInMemoryToFile_and_OfflineEmulation() {
+  const std::string tmpdir = FileManager::GetDirectoryFromFilePath(GenerateTemporaryFileName());
+  CleanUpQueueFiles(tmpdir);
+  ScopedRemoveFile remover(tmpdir + alohalytics::kCurrentFileName);
+  MessagesQueue q;
+  std::string archived_file, second_archived_file;
+  {
+    q.PushMessage(kTestMessage);    // This one goes into the memory storage.
+    q.SetStorageDirectory(tmpdir);  // Here message shoud move from memory into the file.
+    std::thread worker([&q]() { q.PushMessage(kTestWorkerMessage); });
+    worker.join();
+
+    // Wait until messages will be stored into the file.
+    // NOTE: THIS IS NOT A PRODUCTION-READY PRACTICE! NEVER USE IT IN PRODUCTION!
+    // Here it is used for tests simplification only.
```

Comment:
No, file is written in another thread context (WorkerThread in the queue) so this trick does not work.